pytorch_lowerprecision.md

```diff
@@ -0,0 +1,246 @@
+# Proposal of sub-byte low precision inference for PyTorch/IPEX
+
+**Authors:**
+* @xinhe3, hengyume
+
+## **Summary**
+In this proposal, we mainly discuss the storage data type, compute data type, quantization, and serialization, to facilitate the low-bit (INT4/FP4) LLM inference for PyTorch/IPEX.
+
+## **Motivation**
+The community is active to support sub-byte low precision inference for LLMs, where the sub-byte may include INT4, FP4, NF4, INT2/3/5/6/7, MX formats, etc. The ecosystem is bringing up new idea of group-wise sub-byte support from the recent work like [llama.cpp](https://github.com/ggerganov/llama.cpp) and [OCP MX format](https://www.opencompute.org/blog/amd-arm-intel-meta-microsoft-nvidia-and-qualcomm-standardize-next-generation-narrow-precision-data-formats-for-ai).
+
+## **Storage data type**
+
+### Option 1: reuse torch.uint8 
+We follow the existing storage data type torch.uint8 to represent INT4/FP4 and there is no new storage required. In particular, uint8 is interchangeable with a pair of uint4 in PyTorch.
+
+```cpp
+struct bit4x2 {
+  int8_t x : 4;
+  int8_t y : 4;
+  bit4x2(int8_t v) : x(v), y(v) {}
+  bit4x2() : x(0), y(0) {}
+};
+
+struct int4x2 : bit4x2 {
+    int4x2(int8_t v) : bit4x2(v) {}
+    int4x2() : bit4x2() {}
+    static int8_t convert(int8_t src) {
+        int32_t dst = src;
+        dst = dst >= 0 ? dst + 8 : dst - 8;
+        dst = dst / 16;
+        dst = dst > 7 ? 7 : dst;
+        dst = dst < -8 ? -8 : dst;
+        return static_cast<int8_t>(dst);
+    }
+}
+```
+
+```python
+x_uint8 = torch.tensor([1, 7]).to(torch.uint8)
+x_uint4 = x_uint8.to(torch.uint4x2)
+# x_uint4: 00011111/b
+
+scales = torch.rand(1)
+zero_points = torch.rand(1)
+
+woqlinear = nn.WOQLinear(group_size=2) # new op required
+output = woqlinear(x_uint4, scales, zero_points) 
```

Comment:
the first input of woqlinear should be fp32 but not uint4. uint4 should be only used in weight part. 

the remaining codes have same issues

Reply:
nice catch, uint4 tensor should be parameters of woqlinear not input, fixed in [16c286a](https://github.com/intel/neural-speed/pull/15/commits/16c286a4cf1724c67adb8fdb89c99aa73c246a2e)